# 🐍 Pandas with Pokemon Dataset

<center><img src="https://www.fayerwayer.com/resizer/3gNBYtOhENu4KWGdpWHORkjAMhQ=/800x0/filters:format(jpg):quality(70)/cloudfront-us-east-1.images.arcpublishing.com/metroworldnews/ZOP2OEAULVFUJJIZCFHWIZP5XM.jpeg" width="1000px" align="center" /></center>




## 👨‍🏫 Intro to pandas

Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language.

### What kind of data does pandas handle?

When working with tabular data, such as data stored in spreadsheets or databases, pandas is the right tool for you. pandas will help you to explore, clean, and process your data. In pandas, a data table is called a DataFrame.

<center><img src="https://pandas.pydata.org/docs/_images/01_table_dataframe.svg" /></center>

### Comparison with spreadsheets

General terminology translation

| pandas |   Excel |  
| ----------- | ----------- |
| DataFrame | worksheet |
| Series | column |
| Index | row headings |
| row | row |
| NaN | empty cell |


### How do I read and write tabular data?

Pandas supports the integration with many file formats or data sources out of the box (csv, excel, sql, json, parquet,…). Importing data from each of these data sources is provided by function with the prefix `read_*`. Similarly, the `to_*` methods are used to store data.

<center><img src="https://pandas.pydata.org/docs/_images/02_io_readwrite.svg" /></center>


## 👨‍🏭 Installation

Pandas can be installed via pip from PyPI: `pip install pandas`


Working with conda?

Pandas is part of the Anaconda distribution and can be installed with Anaconda or Miniconda:
`conda install pandas`

## 📚 About the Dataset

### Context

This dataset contains information on all 802 Pokemon from all Seven Generations of Pokemon. The information contained in this dataset include Base Stats, Performance against Other Types, Height, Weight, Classification, Egg Steps, Experience Points, Abilities, etc. The information was scraped from http://serebii.net/

### Content
- **name**: The English name of the Pokemon
- **japanese_name**: The Original Japanese name of the Pokemon
- **pokedex_number**: The entry number of the Pokemon in the National Pokedex
- **percentage_male**: The percentage of the species that are male. Blank if the Pokemon is genderless.
- **type1**: The Primary Type of the Pokemon
- **type2**: The Secondary Type of the Pokemon
- **classification**: The Classification of the Pokemon as described by the Sun and Moon Pokedex
- **height_m**: Height of the Pokemon in metres
- **weight_kg**: The Weight of the Pokemon in kilograms
- **capture_rate**: Capture Rate of the Pokemon
- **baseeggsteps**: The number of steps required to hatch an egg of the Pokemon
- **abilities**: A stringified list of abilities that the Pokemon is capable of having
- **experience_growth**: The Experience Growth of the Pokemon
- **base_happiness**: Base Happiness of the Pokemon
- **against_?**: Eighteen features that denote the amount of damage taken against an attack of a particular type
- **hp**: The Base HP of the Pokemon
- **attack**: The Base Attack of the Pokemon
- **defense**: The Base Defense of the Pokemon
- **sp_attack**: The Base Special Attack of the Pokemon
- **sp_defense**: The Base Special Defense of the Pokemon
- **speed**: The Base Speed of the Pokemon
- **generation**: The numbered generation which the Pokemon was first introduced
- **is_legendary**: Denotes if the Pokemon is legendary.

## 👨‍💻 The Code

### Importing libraries

In [ ]:
!pip install pandas

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import requests
import pandas as pd
import matplotlib.pyplot as plt

### Basic setup

In [ ]:
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [ ]:
pd.__version__

'1.3.5'

### Creating a Pandas Dataframe from a Dictionary

In [ ]:
# dictionary with list object in values
details = {
    'Name' : ['Bulbasaur', 'Charmander', 'Squirtle'],
    'Generation' : [1, 1, 1],
    'Type' : ['grass', 'fire', 'water'],
}
  
# creating a Dataframe object 
df = pd.DataFrame(details)
  
df

Name  Generation   Type
0   Bulbasaur           1  grass
1  Charmander           1   fire
2    Squirtle           1  water

### Downloading data for this lab

Fetching file

In [ ]:
dataset_url="https://github.com/DataEngineering-LATAM/Python-StudyClub/raw/17684af6ae16e137e399acee33d6dca7493731a2/Worshop%20Python%20%23%209:%20Pandas/pokemon.zip"
response = requests.get(url=dataset_url, allow_redirects=True)

# Get response status
response.status_code

#view the contents
#response.content

200

In [ ]:
type(response.content)

bytes

Writing to local disk

In [ ]:
with open("pokemon.zip", "wb") as binary_file:
   
    # Write bytes to file
    binary_file.write(response.content)

### Importing data from local disk

In [ ]:
df = pd.read_csv(filepath_or_buffer='pokemon.zip', compression='zip', header=0, sep=',', quotechar='"')
#pd.read_excel()
#pd.read_table()
#pd.read_sql()
#pd.read_json()
#pd.read_html()

In [ ]:
type(df)

pandas.core.frame.DataFrame

### Importing data directly from the source

In [ ]:
df = pd.read_csv(filepath_or_buffer=dataset_url, compression='zip', header=0, sep=',', quotechar='"')
type(df)

pandas.core.frame.DataFrame

### Return a tuple representing the dimensionality of the DataFrame.



In [ ]:
df.shape

(801, 41)

This dataset contains:

- number of rows: 801
- number of columns: 41

### Print a concise summary of a DataFrame.



In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Data columns (total 41 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   abilities          801 non-null    object 
 1   against_bug        801 non-null    float64
 2   against_dark       801 non-null    float64
 3   against_dragon     801 non-null    float64
 4   against_electric   801 non-null    float64
 5   against_fairy      801 non-null    float64
 6   against_fight      801 non-null    float64
 7   against_fire       801 non-null    float64
 8   against_flying     801 non-null    float64
 9   against_ghost      801 non-null    float64
 10  against_grass      801 non-null    float64
 11  against_ground     801 non-null    float64
 12  against_ice        801 non-null    float64
 13  against_normal     801 non-null    float64
 14  against_poison     801 non-null    float64
 15  against_psychic    801 non-null    float64
 16  against_rock       801 non

### Return the first n rows.

In [ ]:
df.head(n=5)

abilities  against_bug  against_dark  against_dragon  \
0  ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
1  ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
2  ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
3     ['Blaze', 'Solar Power']        0.500         1.000           1.000   
4     ['Blaze', 'Solar Power']        0.500         1.000           1.000   

   against_electric  against_fairy  against_fight  against_fire  \
0             0.500          0.500          0.500         2.000   
1             0.500          0.500          0.500         2.000   
2             0.500          0.500          0.500         2.000   
3             1.000          0.500          1.000         0.500   
4             1.000          0.500          1.000         0.500   

   against_flying  against_ghost  against_grass  against_ground  against_ice  \
0           2.000          1.000          0.250           1.000        2.000   
1           2.000          1.000          0.250           1.000        2.000   
2           2.000          1.000          0.250           1.000        2.000   
3           1.000          1.000          0.500           2.000        0.500   
4           1.000          1.000          0.500           2.000        0.500   

   against_normal  against_poison  against_psychic  against_rock  \
0           1.000           1.000            2.000         1.000   
1           1.000           1.000            2.000         1.000   
2           1.000           1.000            2.000         1.000   
3           1.000           1.000            1.000         2.000   
4           1.000           1.000            1.000         2.000   

   against_steel  against_water  attack  base_egg_steps  base_happiness  \
0          1.000          0.500      49            5120              70   
1          1.000          0.500      62            5120              70   
2          1.000          0.500     100            5120              70   
3          0.500          2.000      52            5120              70   
4          0.500          2.000      64            5120              70   

   base_total capture_rate   classfication  defense  experience_growth  \
0         318           45    Seed Pokémon       49            1059860   
1         405           45    Seed Pokémon       63            1059860   
2         625           45    Seed Pokémon      123            1059860   
3         309           45  Lizard Pokémon       43            1059860   
4         405           45   Flame Pokémon       58            1059860   

   height_m  hp     japanese_name        name  percentage_male  \
0     0.700  45  Fushigidaneフシギダネ   Bulbasaur           88.100   
1     1.000  60   Fushigisouフシギソウ     Ivysaur           88.100   
2     2.000  80  Fushigibanaフシギバナ    Venusaur           88.100   
3     0.600  39      Hitokageヒトカゲ  Charmander           88.100   
4     1.100  58       Lizardoリザード  Charmeleon           88.100   

   pokedex_number  sp_attack  sp_defense  speed  type1   type2  weight_kg  \
0               1         65          65     45  grass  poison      6.900   
1               2         80          80     60  grass  poison     13.000   
2               3        122         120     80  grass  poison    100.000   
3               4         60          50     65   fire     NaN      8.500   
4               5         80          65     80   fire     NaN     19.000   

   generation  is_legendary  
0           1             0  
1           1             0  
2           1             0  
3           1             0  
4           1             0

### Return the last n rows.

In [ ]:
df.tail(n=5)

abilities  against_bug  against_dark  against_dragon  \
796  ['Beast Boost']        0.250         1.000           0.500   
797  ['Beast Boost']        1.000         1.000           0.500   
798  ['Beast Boost']        2.000         0.500           2.000   
799  ['Prism Armor']        2.000         2.000           1.000   
800   ['Soul-Heart']        0.250         0.500           0.000   

     against_electric  against_fairy  against_fight  against_fire  \
796             2.000          0.500          1.000         2.000   
797             0.500          0.500          2.000         4.000   
798             0.500          4.000          2.000         0.500   
799             1.000          1.000          0.500         1.000   
800             1.000          0.500          1.000         2.000   

     against_flying  against_ghost  against_grass  against_ground  \
796           0.500          1.000          0.250           0.000   
797           1.000          1.000          0.250           1.000   
798           1.000          0.500          0.500           1.000   
799           1.000          2.000          1.000           1.000   
800           0.500          1.000          0.500           2.000   

     against_ice  against_normal  against_poison  against_psychic  \
796        1.000           0.500           0.000            0.500   
797        1.000           0.500           0.000            0.500   
798        2.000           1.000           1.000            0.000   
799        1.000           1.000           1.000            0.500   
800        0.500           0.500           0.000            0.500   

     against_rock  against_steel  against_water  attack  base_egg_steps  \
796         1.000          0.500          1.000     101           30720   
797         0.500          0.500          0.500     181           30720   
798         1.000          1.000          0.500     101           30720   
799         1.000          1.000          1.000     107           30720   
800         0.500          1.000          1.000      95           30720   

     base_happiness  base_total capture_rate        classfication  defense  \
796               0         570           25       Launch Pokémon      103   
797               0         570          255  Drawn Sword Pokémon      131   
798               0         570           15    Junkivore Pokémon       53   
799               0         600            3        Prism Pokémon      101   
800               0         600            3   Artificial Pokémon      115   

     experience_growth  height_m   hp    japanese_name        name  \
796            1250000     9.200   97   Tekkaguyaテッカグヤ  Celesteela   
797            1250000     0.300   59  Kamiturugiカミツルギ     Kartana   
798            1250000     5.500  223  Akuzikingアクジキング    Guzzlord   
799            1250000     2.400   97    Necrozmaネクロズマ    Necrozma   
800            1250000     1.000   80     Magearnaマギアナ    Magearna   

     percentage_male  pokedex_number  sp_attack  sp_defense  speed    type1  \
796              NaN             797        107         101     61    steel   
797              NaN             798         59          31    109    grass   
798              NaN             799         97          53     43     dark   
799              NaN             800        127          89     79  psychic   
800              NaN             801        130         115     65    steel   

      type2  weight_kg  generation  is_legendary  
796  flying    999.900           7             1  
797   steel      0.100           7             1  
798  dragon    888.000           7             1  
799     NaN    230.000           7             1  
800   fairy     80.500           7             1

### Renaming columns

Let's fix the **'classfication'** column name

In [ ]:
df.rename(columns={"classfication": "classification"})

abilities  against_bug  against_dark  against_dragon  \
0    ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
1    ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
2    ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
3       ['Blaze', 'Solar Power']        0.500         1.000           1.000   
4       ['Blaze', 'Solar Power']        0.500         1.000           1.000   
..                           ...          ...           ...             ...   
796              ['Beast Boost']        0.250         1.000           0.500   
797              ['Beast Boost']        1.000         1.000           0.500   
798              ['Beast Boost']        2.000         0.500           2.000   
799              ['Prism Armor']        2.000         2.000           1.000   
800               ['Soul-Heart']        0.250         0.500           0.000   

     against_electric  against_fairy  against_fight  against_fire  \
0               0.500          0.500          0.500         2.000   
1               0.500          0.500          0.500         2.000   
2               0.500          0.500          0.500         2.000   
3               1.000          0.500          1.000         0.500   
4               1.000          0.500          1.000         0.500   
..                ...            ...            ...           ...   
796             2.000          0.500          1.000         2.000   
797             0.500          0.500          2.000         4.000   
798             0.500          4.000          2.000         0.500   
799             1.000          1.000          0.500         1.000   
800             1.000          0.500          1.000         2.000   

     against_flying  against_ghost  against_grass  against_ground  \
0             2.000          1.000          0.250           1.000   
1             2.000          1.000          0.250           1.000   
2             2.000          1.000          0.250           1.000   
3             1.000          1.000          0.500           2.000   
4             1.000          1.000          0.500           2.000   
..              ...            ...            ...             ...   
796           0.500          1.000          0.250           0.000   
797           1.000          1.000          0.250           1.000   
798           1.000          0.500          0.500           1.000   
799           1.000          2.000          1.000           1.000   
800           0.500          1.000          0.500           2.000   

     against_ice  against_normal  against_poison  against_psychic  \
0          2.000           1.000           1.000            2.000   
1          2.000           1.000           1.000            2.000   
2          2.000           1.000           1.000            2.000   
3          0.500           1.000           1.000            1.000   
4          0.500           1.000           1.000            1.000   
..           ...             ...             ...              ...   
796        1.000           0.500           0.000            0.500   
797        1.000           0.500           0.000            0.500   
798        2.000           1.000           1.000            0.000   
799        1.000           1.000           1.000            0.500   
800        0.500           0.500           0.000            0.500   

     against_rock  against_steel  against_water  attack  base_egg_steps  \
0           1.000          1.000          0.500      49            5120   
1           1.000          1.000          0.500      62            5120   
2           1.000          1.000          0.500     100            5120   
3           2.000          0.500          2.000      52            5120   
4           2.000          0.500          2.000      64            5120   
..            ...            ...            ...     ...             ...   
796         1.000          0.500          1.000     101       

### Let's reorder the columns

In [ ]:
df.insert(0, 'generation', df.pop('generation'))
df.insert(1, 'is_legendary', df.pop('is_legendary'))
df.insert(2, 'type1', df.pop('type1'))
df.insert(3, 'type2', df.pop('type2'))
df.insert(4, 'classification', df.pop('classfication')) # You could have corrected that name here!
df.insert(5, 'name', df.pop('name'))
df.insert(6, 'japanese_name', df.pop('japanese_name'))
df.head(n=10)

generation  is_legendary  type1   type2       classification        name  \
0           1             0  grass  poison         Seed Pokémon   Bulbasaur   
1           1             0  grass  poison         Seed Pokémon     Ivysaur   
2           1             0  grass  poison         Seed Pokémon    Venusaur   
3           1             0   fire     NaN       Lizard Pokémon  Charmander   
4           1             0   fire     NaN        Flame Pokémon  Charmeleon   
5           1             0   fire  flying        Flame Pokémon   Charizard   
6           1             0  water     NaN  Tiny Turtle Pokémon    Squirtle   
7           1             0  water     NaN       Turtle Pokémon   Wartortle   
8           1             0  water     NaN    Shellfish Pokémon   Blastoise   
9           1             0    bug     NaN         Worm Pokémon    Caterpie   

      japanese_name                    abilities  against_bug  against_dark  \
0  Fushigidaneフシギダネ  ['Overgrow', 'Chlorophyll']        1.000         1.000   
1   Fushigisouフシギソウ  ['Overgrow', 'Chlorophyll']        1.000         1.000   
2  Fushigibanaフシギバナ  ['Overgrow', 'Chlorophyll']        1.000         1.000   
3      Hitokageヒトカゲ     ['Blaze', 'Solar Power']        0.500         1.000   
4       Lizardoリザード     ['Blaze', 'Solar Power']        0.500         1.000   
5     Lizardonリザードン     ['Blaze', 'Solar Power']        0.250         1.000   
6      Zenigameゼニガメ     ['Torrent', 'Rain Dish']        1.000         1.000   
7        Kameilカメール     ['Torrent', 'Rain Dish']        1.000         1.000   
8        Kamexカメックス     ['Torrent', 'Rain Dish']        1.000         1.000   
9     Caterpieキャタピー  ['Shield Dust', 'Run Away']        1.000         1.000   

   against_dragon  against_electric  against_fairy  against_fight  \
0           1.000             0.500          0.500          0.500   
1           1.000             0.500          0.500          0.500   
2           1.000             0.500          0.500          0.500   
3           1.000             1.000          0.500          1.000   
4           1.000             1.000          0.500          1.000   
5           1.000             2.000          0.500          0.500   
6           1.000             2.000          1.000          1.000   
7           1.000             2.000          1.000          1.000   
8           1.000             2.000          1.000          1.000   
9           1.000             1.000          1.000          0.500   

   against_fire  against_flying  against_ghost  against_grass  against_ground  \
0         2.000           2.000          1.000          0.250           1.000   
1         2.000           2.000          1.000          0.250           1.000   
2         2.000           2.000          1.000          0.250           1.000   
3         0.500           1.000          1.000          0.500           2.000   
4         0.500           1.000          1.000          0.500           2.000   
5         0.500           1.000          1.000          0.250           0.000   
6         0.500           1.000          1.000          2.000           1.000   
7         0.500           1.000          1.000          2.000           1.000   
8         0.500           1.000          1.000          2.000           1.000   
9         2.000           2.000          1.000          0.500           0.500   

   against_ice  against_normal  against_poison  against_psychic  against_rock  \
0        2.000           1.000           1.000            2.000         1.000   
1        2.000           1.000           1.000            2.000         1.000   
2        2.000           1.000           1.000            2.000         1.000   
3        0.500           1.000           1.000            1.000         2.000   
4        0.500           1.000           1.000            1.000         2.000   
5        1.000           1.000           1.000            1.000         4.000   
6        0.500           1.000           1.0

### How to drop a column?

In [ ]:
df.drop('japanese_name', inplace=True, axis=1)
df.head()

generation  is_legendary  type1   type2  classification        name  \
0           1             0  grass  poison    Seed Pokémon   Bulbasaur   
1           1             0  grass  poison    Seed Pokémon     Ivysaur   
2           1             0  grass  poison    Seed Pokémon    Venusaur   
3           1             0   fire     NaN  Lizard Pokémon  Charmander   
4           1             0   fire     NaN   Flame Pokémon  Charmeleon   

                     abilities  against_bug  against_dark  against_dragon  \
0  ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
1  ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
2  ['Overgrow', 'Chlorophyll']        1.000         1.000           1.000   
3     ['Blaze', 'Solar Power']        0.500         1.000           1.000   
4     ['Blaze', 'Solar Power']        0.500         1.000           1.000   

   against_electric  against_fairy  against_fight  against_fire  \
0             0.500          0.500          0.500         2.000   
1             0.500          0.500          0.500         2.000   
2             0.500          0.500          0.500         2.000   
3             1.000          0.500          1.000         0.500   
4             1.000          0.500          1.000         0.500   

   against_flying  against_ghost  against_grass  against_ground  against_ice  \
0           2.000          1.000          0.250           1.000        2.000   
1           2.000          1.000          0.250           1.000        2.000   
2           2.000          1.000          0.250           1.000        2.000   
3           1.000          1.000          0.500           2.000        0.500   
4           1.000          1.000          0.500           2.000        0.500   

   against_normal  against_poison  against_psychic  against_rock  \
0           1.000           1.000            2.000         1.000   
1           1.000           1.000            2.000         1.000   
2           1.000           1.000            2.000         1.000   
3           1.000           1.000            1.000         2.000   
4           1.000           1.000            1.000         2.000   

   against_steel  against_water  attack  base_egg_steps  base_happiness  \
0          1.000          0.500      49            5120              70   
1          1.000          0.500      62            5120              70   
2          1.000          0.500     100            5120              70   
3          0.500          2.000      52            5120              70   
4          0.500          2.000      64            5120              70   

   base_total capture_rate  defense  experience_growth  height_m  hp  \
0         318           45       49            1059860     0.700  45   
1         405           45       63            1059860     1.000  60   
2         625           45      123            1059860     2.000  80   
3         309           45       43            1059860     0.600  39   
4         405           45       58            1059860     1.100  58   

   percentage_male  pokedex_number  sp_attack  sp_defense  speed  weight_kg  
0           88.100               1         65          65     45      6.900  
1           88.100               2         80          80     60     13.000  
2           88.100               3        122         120     80    100.000  
3           88.100               4         60          50     65      8.500  
4           88.100               5         80          65     80     19.000

### How do I change the index of a DataFrame?

In [ ]:
df.index=df['name']
df.head(3)

generation  is_legendary  type1   type2 classification       name  \
name                                                                           
Bulbasaur           1             0  grass  poison   Seed Pokémon  Bulbasaur   
Ivysaur             1             0  grass  poison   Seed Pokémon    Ivysaur   
Venusaur            1             0  grass  poison   Seed Pokémon   Venusaur   

                             abilities  against_bug  against_dark  \
name                                                                
Bulbasaur  ['Overgrow', 'Chlorophyll']        1.000         1.000   
Ivysaur    ['Overgrow', 'Chlorophyll']        1.000         1.000   
Venusaur   ['Overgrow', 'Chlorophyll']        1.000         1.000   

           against_dragon  against_electric  against_fairy  against_fight  \
name                                                                        
Bulbasaur           1.000             0.500          0.500          0.500   
Ivysaur             1.000             0.500          0.500          0.500   
Venusaur            1.000             0.500          0.500          0.500   

           against_fire  against_flying  against_ghost  against_grass  \
name                                                                    
Bulbasaur         2.000           2.000          1.000          0.250   
Ivysaur           2.000           2.000          1.000          0.250   
Venusaur          2.000           2.000          1.000          0.250   

           against_ground  against_ice  against_normal  against_poison  \
name                                                                     
Bulbasaur           1.000        2.000           1.000           1.000   
Ivysaur             1.000        2.000           1.000           1.000   
Venusaur            1.000        2.000           1.000           1.000   

           against_psychic  against_rock  against_steel  against_water  \
name                                                                     
Bulbasaur            2.000         1.000          1.000          0.500   
Ivysaur              2.000         1.000          1.000          0.500   
Venusaur             2.000         1.000          1.000          0.500   

           attack  base_egg_steps  base_happiness  base_total capture_rate  \
name                                                                         
Bulbasaur      49            5120              70         318           45   
Ivysaur        62            5120              70         405           45   
Venusaur      100            5120              70         625           45   

           defense  experience_growth  height_m  hp  percentage_male  \
name                                                                   
Bulbasaur       49            1059860     0.700  45           88.100   
Ivysaur         63            1059860     1.000  60           88.100   
Venusaur       123            1059860     2.000  80           88.100   

           pokedex_number  sp_attack  sp_defense  speed  weight_kg  
name                                                                
Bulbasaur               1         65          65     45      6.900  
Ivysaur                 2         80          80     60     13.000  
Venusaur                3        122         120     80    100.000

### How do I select a subset of a table?

Selecting or filtering specific rows and/or columns? Filtering the data on a condition? Methods for slicing, selecting, and extracting the data you need are available in pandas.

<img src="https://pandas.pydata.org/docs/_images/03_subset_columns_rows.svg" />

#### How do I select specific columns from a DataFrame?


In [ ]:
df['name']

name
Bulbasaur      Bulbasaur
Ivysaur          Ivysaur
Venusaur        Venusaur
Charmander    Charmander
Charmeleon    Charmeleon
                 ...    
Celesteela    Celesteela
Kartana          Kartana
Guzzlord        Guzzlord
Necrozma        Necrozma
Magearna        Magearna
Name: name, Length: 801, dtype: object

In [ ]:
some_columns=['generation','classification','type1','type2','name']
df[some_columns]

generation       classification    type1   type2        name
name                                                                    
Bulbasaur            1         Seed Pokémon    grass  poison   Bulbasaur
Ivysaur              1         Seed Pokémon    grass  poison     Ivysaur
Venusaur             1         Seed Pokémon    grass  poison    Venusaur
Charmander           1       Lizard Pokémon     fire     NaN  Charmander
Charmeleon           1        Flame Pokémon     fire     NaN  Charmeleon
...                ...                  ...      ...     ...         ...
Celesteela           7       Launch Pokémon    steel  flying  Celesteela
Kartana              7  Drawn Sword Pokémon    grass   steel     Kartana
Guzzlord             7    Junkivore Pokémon     dark  dragon    Guzzlord
Necrozma             7        Prism Pokémon  psychic     NaN    Necrozma
Magearna             7   Artificial Pokémon    steel   fairy    Magearna

[801 rows x 5 columns]

In [ ]:
filter = (df['generation']==1)  & (df['type1']=='psychic')
df[ filter ]

generation  is_legendary    type1  type2       classification  \
name                                                                      
Abra               1             0  psychic    NaN          Psi Pokémon   
Kadabra            1             0  psychic    NaN          Psi Pokémon   
Alakazam           1             0  psychic    NaN          Psi Pokémon   
Drowzee            1             0  psychic    NaN     Hypnosis Pokémon   
Hypno              1             0  psychic    NaN     Hypnosis Pokémon   
Mr. Mime           1             0  psychic  fairy      Barrier Pokémon   
Mewtwo             1             1  psychic    NaN      Genetic Pokémon   
Mew                1             1  psychic    NaN  New Species Pokémon   

              name                                      abilities  \
name                                                                
Abra          Abra  ['Synchronize', 'Inner Focus', 'Magic Guard']   
Kadabra    Kadabra  ['Synchronize', 'Inner Focus', 'Magic Guard']   
Alakazam  Alakazam  ['Synchronize', 'Inner Focus', 'Magic Guard']   
Drowzee    Drowzee        ['Insomnia', 'Forewarn', 'Inner Focus']   
Hypno        Hypno        ['Insomnia', 'Forewarn', 'Inner Focus']   
Mr. Mime  Mr. Mime         ['Soundproof', 'Filter', 'Technician']   
Mewtwo      Mewtwo                        ['Pressure', 'Unnerve']   
Mew            Mew                                ['Synchronize']   

          against_bug  against_dark  against_dragon  against_electric  \
name                                                                    
Abra            2.000         2.000           1.000             1.000   
Kadabra         2.000         2.000           1.000             1.000   
Alakazam        2.000         2.000           1.000             1.000   
Drowzee         2.000         2.000           1.000             1.000   
Hypno           2.000         2.000           1.000             1.000   
Mr. Mime        1.000         1.000           0.000             1.000   
Mewtwo          2.000         2.000           1.000             1.000   
Mew             2.000         2.000           1.000             1.000   

          against_fairy  against_fight  against_fire  against_flying  \
name                                                                   
Abra              1.000          0.500         1.000           1.000   
Kadabra           1.000          0.500         1.000           1.000   
Alakazam          1.000          0.500         1.000           1.000   
Drowzee           1.000          0.500         1.000           1.000   
Hypno             1.000          0.500         1.000           1.000   
Mr. Mime          1.000          0.250         1.000           1.000   
Mewtwo            1.000          0.500         1.000           1.000   
Mew               1.000          0.500         1.000           1.000   

          against_ghost  against_grass  against_ground  against_ice  \
name                                                                  
Abra              2.000          1.000           1.000        1.000   
Kadabra           2.000          1.000           1.000        1.000   
Alakazam          2.000          1.000           1.000        1.000   
Drowzee           2.000          1.000           1.000        1.000   
Hypno             2.000          1.000           1.000        1.000   
Mr. Mime          2.000          1.000           1.000        1.000   
Mewtwo            2.000          1.000           1.000        1.000   
Mew               2.000          1.000           1.000        1.000   

          against_normal  against_poison  against_psychic  against_rock  \
name                                                                      
Abra               1.000           1.000            0.500         1.000   
Kadabra            1.000           1.000            0.500         1.000   
Alakazam           1.000           1.000            0.500         1.000   
Drowzee            1.000      

#### How do I select specific rows from a DataFrame?

##### Access a group of rows and columns by label(s)

`.loc[]` is primarily label based, but may also be used with a boolean array.

In [ ]:
df.loc['Bulbasaur']

generation                                     1
is_legendary                                   0
type1                                      grass
type2                                     poison
classification                      Seed Pokémon
name                                   Bulbasaur
abilities            ['Overgrow', 'Chlorophyll']
against_bug                                1.000
against_dark                               1.000
against_dragon                             1.000
against_electric                           0.500
against_fairy                              0.500
against_fight                              0.500
against_fire                               2.000
against_flying                             2.000
against_ghost                              1.000
against_grass                              0.250
against_ground                             1.000
against_ice                                2.000
against_normal                             1.000
against_poison      

In [ ]:
some_indexes=['Bulbasaur','Ivysaur','Venusaur']
df.loc[some_indexes]

generation  is_legendary  type1   type2 classification       name  \
name                                                                           
Bulbasaur           1             0  grass  poison   Seed Pokémon  Bulbasaur   
Ivysaur             1             0  grass  poison   Seed Pokémon    Ivysaur   
Venusaur            1             0  grass  poison   Seed Pokémon   Venusaur   

                             abilities  against_bug  against_dark  \
name                                                                
Bulbasaur  ['Overgrow', 'Chlorophyll']        1.000         1.000   
Ivysaur    ['Overgrow', 'Chlorophyll']        1.000         1.000   
Venusaur   ['Overgrow', 'Chlorophyll']        1.000         1.000   

           against_dragon  against_electric  against_fairy  against_fight  \
name                                                                        
Bulbasaur           1.000             0.500          0.500          0.500   
Ivysaur             1.000             0.500          0.500          0.500   
Venusaur            1.000             0.500          0.500          0.500   

           against_fire  against_flying  against_ghost  against_grass  \
name                                                                    
Bulbasaur         2.000           2.000          1.000          0.250   
Ivysaur           2.000           2.000          1.000          0.250   
Venusaur          2.000           2.000          1.000          0.250   

           against_ground  against_ice  against_normal  against_poison  \
name                                                                     
Bulbasaur           1.000        2.000           1.000           1.000   
Ivysaur             1.000        2.000           1.000           1.000   
Venusaur            1.000        2.000           1.000           1.000   

           against_psychic  against_rock  against_steel  against_water  \
name                                                                     
Bulbasaur            2.000         1.000          1.000          0.500   
Ivysaur              2.000         1.000          1.000          0.500   
Venusaur             2.000         1.000          1.000          0.500   

           attack  base_egg_steps  base_happiness  base_total capture_rate  \
name                                                                         
Bulbasaur      49            5120              70         318           45   
Ivysaur        62            5120              70         405           45   
Venusaur      100            5120              70         625           45   

           defense  experience_growth  height_m  hp  percentage_male  \
name                                                                   
Bulbasaur       49            1059860     0.700  45           88.100   
Ivysaur         63            1059860     1.000  60           88.100   
Venusaur       123            1059860     2.000  80           88.100   

           pokedex_number  sp_attack  sp_defense  speed  weight_kg  
name                                                                
Bulbasaur               1         65          65     45      6.900  
Ivysaur                 2         80          80     60     13.000  
Venusaur                3        122         120     80    100.000

In [ ]:
try:
  df.loc[0]
except KeyError as e:
  print("This key was not found in the index of the Dataframe:",e)

This key was not found in the index of the Dataframe: 0


##### Access a group of rows and columns by integer-location based indexing

`.iloc[]` is primarily integer position based (from `0` to `length-1` of the axis), but may also be used with a boolean array.

In [ ]:
df.iloc[0]

generation                                     1
is_legendary                                   0
type1                                      grass
type2                                     poison
classification                      Seed Pokémon
name                                   Bulbasaur
abilities            ['Overgrow', 'Chlorophyll']
against_bug                                1.000
against_dark                               1.000
against_dragon                             1.000
against_electric                           0.500
against_fairy                              0.500
against_fight                              0.500
against_fire                               2.000
against_flying                             2.000
against_ghost                              1.000
against_grass                              0.250
against_ground                             1.000
against_ice                                2.000
against_normal                             1.000
against_poison      

In [ ]:
some_integers = [0,1,2]
df.iloc[some_integers]

generation  is_legendary  type1   type2 classification       name  \
name                                                                           
Bulbasaur           1             0  grass  poison   Seed Pokémon  Bulbasaur   
Ivysaur             1             0  grass  poison   Seed Pokémon    Ivysaur   
Venusaur            1             0  grass  poison   Seed Pokémon   Venusaur   

                             abilities  against_bug  against_dark  \
name                                                                
Bulbasaur  ['Overgrow', 'Chlorophyll']        1.000         1.000   
Ivysaur    ['Overgrow', 'Chlorophyll']        1.000         1.000   
Venusaur   ['Overgrow', 'Chlorophyll']        1.000         1.000   

           against_dragon  against_electric  against_fairy  against_fight  \
name                                                                        
Bulbasaur           1.000             0.500          0.500          0.500   
Ivysaur             1.000             0.500          0.500          0.500   
Venusaur            1.000             0.500          0.500          0.500   

           against_fire  against_flying  against_ghost  against_grass  \
name                                                                    
Bulbasaur         2.000           2.000          1.000          0.250   
Ivysaur           2.000           2.000          1.000          0.250   
Venusaur          2.000           2.000          1.000          0.250   

           against_ground  against_ice  against_normal  against_poison  \
name                                                                     
Bulbasaur           1.000        2.000           1.000           1.000   
Ivysaur             1.000        2.000           1.000           1.000   
Venusaur            1.000        2.000           1.000           1.000   

           against_psychic  against_rock  against_steel  against_water  \
name                                                                     
Bulbasaur            2.000         1.000          1.000          0.500   
Ivysaur              2.000         1.000          1.000          0.500   
Venusaur             2.000         1.000          1.000          0.500   

           attack  base_egg_steps  base_happiness  base_total capture_rate  \
name                                                                         
Bulbasaur      49            5120              70         318           45   
Ivysaur        62            5120              70         405           45   
Venusaur      100            5120              70         625           45   

           defense  experience_growth  height_m  hp  percentage_male  \
name                                                                   
Bulbasaur       49            1059860     0.700  45           88.100   
Ivysaur         63            1059860     1.000  60           88.100   
Venusaur       123            1059860     2.000  80           88.100   

           pokedex_number  sp_attack  sp_defense  speed  weight_kg  
name                                                                
Bulbasaur               1         65          65     45      6.900  
Ivysaur                 2         80          80     60     13.000  
Venusaur                3        122         120     80    100.000

In [ ]:
df.iloc[0:3]

generation  is_legendary  type1   type2 classification       name  \
name                                                                           
Bulbasaur           1             0  grass  poison   Seed Pokémon  Bulbasaur   
Ivysaur             1             0  grass  poison   Seed Pokémon    Ivysaur   
Venusaur            1             0  grass  poison   Seed Pokémon   Venusaur   

                             abilities  against_bug  against_dark  \
name                                                                
Bulbasaur  ['Overgrow', 'Chlorophyll']        1.000         1.000   
Ivysaur    ['Overgrow', 'Chlorophyll']        1.000         1.000   
Venusaur   ['Overgrow', 'Chlorophyll']        1.000         1.000   

           against_dragon  against_electric  against_fairy  against_fight  \
name                                                                        
Bulbasaur           1.000             0.500          0.500          0.500   
Ivysaur             1.000             0.500          0.500          0.500   
Venusaur            1.000             0.500          0.500          0.500   

           against_fire  against_flying  against_ghost  against_grass  \
name                                                                    
Bulbasaur         2.000           2.000          1.000          0.250   
Ivysaur           2.000           2.000          1.000          0.250   
Venusaur          2.000           2.000          1.000          0.250   

           against_ground  against_ice  against_normal  against_poison  \
name                                                                     
Bulbasaur           1.000        2.000           1.000           1.000   
Ivysaur             1.000        2.000           1.000           1.000   
Venusaur            1.000        2.000           1.000           1.000   

           against_psychic  against_rock  against_steel  against_water  \
name                                                                     
Bulbasaur            2.000         1.000          1.000          0.500   
Ivysaur              2.000         1.000          1.000          0.500   
Venusaur             2.000         1.000          1.000          0.500   

           attack  base_egg_steps  base_happiness  base_total capture_rate  \
name                                                                         
Bulbasaur      49            5120              70         318           45   
Ivysaur        62            5120              70         405           45   
Venusaur      100            5120              70         625           45   

           defense  experience_growth  height_m  hp  percentage_male  \
name                                                                   
Bulbasaur       49            1059860     0.700  45           88.100   
Ivysaur         63            1059860     1.000  60           88.100   
Venusaur       123            1059860     2.000  80           88.100   

           pokedex_number  sp_attack  sp_defense  speed  weight_kg  
name                                                                
Bulbasaur               1         65          65     45      6.900  
Ivysaur                 2         80          80     60     13.000  
Venusaur                3        122         120     80    100.000

### Generate descriptive statistics.

Descriptive statistics include those that summarize the *central tendency*, *dispersion* and *shape* of a dataset’s distribution, excluding NaN values.



In [ ]:
df.describe()
#pd.median()
#pd.count()
#pd.std()
#pd.max()
#pd.min()

generation  is_legendary  against_bug  against_dark  against_dragon  \
count     801.000       801.000      801.000       801.000         801.000   
mean        3.690         0.087        0.996         1.057           0.969   
std         1.930         0.283        0.597         0.438           0.353   
min         1.000         0.000        0.250         0.250           0.000   
25%         2.000         0.000        0.500         1.000           1.000   
50%         4.000         0.000        1.000         1.000           1.000   
75%         5.000         0.000        1.000         1.000           1.000   
max         7.000         1.000        4.000         4.000           2.000   

       against_electric  against_fairy  against_fight  against_fire  \
count           801.000        801.000        801.000       801.000   
mean              1.074          1.069          1.066         1.135   
std               0.655          0.522          0.717         0.692   
min               0.000          0.250          0.000         0.250   
25%               0.500          1.000          0.500         0.500   
50%               1.000          1.000          1.000         1.000   
75%               1.000          1.000          1.000         2.000   
max               4.000          4.000          4.000         4.000   

       against_flying  against_ghost  against_grass  against_ground  \
count         801.000        801.000        801.000         801.000   
mean            1.193          0.985          1.034           1.098   
std             0.604          0.558          0.789           0.739   
min             0.250          0.000          0.250           0.000   
25%             1.000          1.000          0.500           1.000   
50%             1.000          1.000          1.000           1.000   
75%             1.000          1.000          1.000           1.000   
max             4.000          4.000          4.000           4.000   

       against_ice  against_normal  against_poison  against_psychic  \
count      801.000         801.000         801.000          801.000   
mean         1.208           0.887           0.975            1.005   
std          0.735           0.266           0.549            0.495   
min          0.250           0.000           0.000            0.000   
25%          0.500           1.000           0.500            1.000   
50%          1.000           1.000           1.000            1.000   
75%          2.000           1.000           1.000            1.000   
max          4.000           1.000           4.000            4.000   

       against_rock  against_steel  against_water  attack  base_egg_steps  \
count       801.000        801.000        801.000 801.000         801.000   
mean          1.250          0.983          1.058  77.858        7191.011   
std           0.697          0.500          0.607  32.159        6558.220   
min           0.250          0.250          0.250   5.000        1280.000   
25%           1.000          0.500          0.500  55.000        5120.000   
50%           1.000          1.000          1.000  75.000        5120.000   
75%           2.000          1.000          1.000 100.000        6400.000   
max           4.000          4.000          4.000 185.000       30720.000   

       base_happiness  base_total  defense  experience_growth  height_m  \
count         801.000     801.000  801.000            801.000   781.000   
mean           65.362     428.377   73.009        1054995.905     1.164   
std            19.599     119.204   30.769         160255.835     1.080   
min             0.000     180.000    5.000         600000.000     0.100   
25%            70.000     320.000   50.000        1000000.000     0.600   
50%            70.000     435.000   70.000        1000000.000     1.000   
75%            70.000     505.000   90.000        1059860.000     1.500   
max           140.000     780.000  230.000        1640000.000    14.500   

           hp 

Generate the mean on all the columns

In [ ]:
df.mean()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


generation                3.690
is_legendary              0.087
against_bug               0.996
against_dark              1.057
against_dragon            0.969
against_electric          1.074
against_fairy             1.069
against_fight             1.066
against_fire              1.135
against_flying            1.193
against_ghost             0.985
against_grass             1.034
against_ground            1.098
against_ice               1.208
against_normal            0.887
against_poison            0.975
against_psychic           1.005
against_rock              1.250
against_steel             0.983
against_water             1.058
attack                   77.858
base_egg_steps         7191.011
base_happiness           65.362
base_total              428.377
defense                  73.009
experience_growth   1054995.905
height_m                  1.164
hp                       68.959
percentage_male          55.156
pokedex_number          401.000
sp_attack                71.306
sp_defen

### How to detect missing values.

Return a boolean same-sized object indicating if the values are NA. NA values, such as None or numpy.NaN, gets mapped to True values. Everything else gets mapped to False values. Characters such as empty strings '' or numpy.inf are not considered NA values (unless you set pandas.options.mode.use_inf_as_na = True).

In [ ]:
df.isna()

generation  is_legendary  type1  type2  classification   name  \
name                                                                        
Bulbasaur        False         False  False  False           False  False   
Ivysaur          False         False  False  False           False  False   
Venusaur         False         False  False  False           False  False   
Charmander       False         False  False   True           False  False   
Charmeleon       False         False  False   True           False  False   
...                ...           ...    ...    ...             ...    ...   
Celesteela       False         False  False  False           False  False   
Kartana          False         False  False  False           False  False   
Guzzlord         False         False  False  False           False  False   
Necrozma         False         False  False   True           False  False   
Magearna         False         False  False  False           False  False   

            abilities  against_bug  against_dark  against_dragon  \
name                                                               
Bulbasaur       False        False         False           False   
Ivysaur         False        False         False           False   
Venusaur        False        False         False           False   
Charmander      False        False         False           False   
Charmeleon      False        False         False           False   
...               ...          ...           ...             ...   
Celesteela      False        False         False           False   
Kartana         False        False         False           False   
Guzzlord        False        False         False           False   
Necrozma        False        False         False           False   
Magearna        False        False         False           False   

            against_electric  against_fairy  against_fight  against_fire  \
name                                                                       
Bulbasaur              False          False          False         False   
Ivysaur                False          False          False         False   
Venusaur               False          False          False         False   
Charmander             False          False          False         False   
Charmeleon             False          False          False         False   
...                      ...            ...            ...           ...   
Celesteela             False          False          False         False   
Kartana                False          False          False         False   
Guzzlord               False          False          False         False   
Necrozma               False          False          False         False   
Magearna               False          False          False         False   

            against_flying  against_ghost  against_grass  against_ground  \
name                                                                       
Bulbasaur            False          False          False           False   
Ivysaur              False          False          False           False   
Venusaur             False          False          False           False   
Charmander           False          False          False           False   
Charmeleon           False          False          False           False   
...                    ...            ...            ...             ...   
Celesteela           False          False          False           False   
Kartana              False          False          False           False   
Guzzlord             False          False          False           False   
Necrozma             False          False          False           False   
Magearna             False          False          False           False   

            against_ice  against_normal  against_poison  against_psychic  \
name                                                                   

### How many NULL values do we have in each column?

In [ ]:
df.isna().sum()

generation             0
is_legendary           0
type1                  0
type2                384
classification         0
name                   0
abilities              0
against_bug            0
against_dark           0
against_dragon         0
against_electric       0
against_fairy          0
against_fight          0
against_fire           0
against_flying         0
against_ghost          0
against_grass          0
against_ground         0
against_ice            0
against_normal         0
against_poison         0
against_psychic        0
against_rock           0
against_steel          0
against_water          0
attack                 0
base_egg_steps         0
base_happiness         0
base_total             0
capture_rate           0
defense                0
experience_growth      0
height_m              20
hp                     0
percentage_male       98
pokedex_number         0
sp_attack              0
sp_defense             0
speed                  0
weight_kg             20


### Let's "double click" on *height_m*

In [ ]:
df_height_m_1 = df[ df['height_m'].isna() == 1 ]
df_height_m_1[ ['height_m','classification','type1','type2','name'] ].head(n=50)

height_m                                     classification  \
name                                                                     
Rattata         NaN                                      Mouse Pokémon   
Raticate        NaN                                      Mouse Pokémon   
Raichu          NaN                                      Mouse Pokémon   
Sandshrew       NaN                                      Mouse Pokémon   
Sandslash       NaN                                      Mouse Pokémon   
Vulpix          NaN                                        Fox Pokémon   
Ninetales       NaN                                        Fox Pokémon   
Diglett         NaN                                       Mole Pokémon   
Dugtrio         NaN                                       Mole Pokémon   
Meowth          NaN                                Scratch Cat Pokémon   
Persian         NaN                                 Classy Cat Pokémon   
Geodude         NaN                                       Rock Pokémon   
Graveler        NaN                                       Rock Pokémon   
Golem           NaN                                    Megaton Pokémon   
Grimer          NaN                                     Sludge Pokémon   
Muk             NaN                                     Sludge Pokémon   
Exeggutor       NaN                                    Coconut Pokémon   
Marowak         NaN                                Bone Keeper Pokémon   
Hoopa           NaN  Mischief Pokémon (Confined)Djinn Pokémonn (Unb...   
Lycanroc        NaN                                       Wolf Pokémon   

              type1     type2       name  
name                                      
Rattata      normal      dark    Rattata  
Raticate     normal      dark   Raticate  
Raichu     electric  electric     Raichu  
Sandshrew    ground       ice  Sandshrew  
Sandslash    ground       ice  Sandslash  
Vulpix         fire       ice     Vulpix  
Ninetales      fire       ice  Ninetales  
Diglett      ground    ground    Diglett  
Dugtrio      ground    ground    Dugtrio  
Meowth       normal      dark     Meowth  
Persian      normal      dark    Persian  
Geodude        rock    ground    Geodude  
Graveler       rock    ground   Graveler  
Golem          rock    ground      Golem  
Grimer       poison    poison     Grimer  
Muk          poison    poison        Muk  
Exeggutor     grass   psychic  Exeggutor  
Marowak      ground      fire    Marowak  
Hoopa       psychic     ghost      Hoopa  
Lycanroc       rock       NaN   Lycanroc

### Let's "double click" on *weight_kg*

In [ ]:
df_weight_kg_1 = df[ df['weight_kg'].isna() == 1 ]
df_weight_kg_1[ ['weight_kg','classification','type1','type2','name'] ].head(n=50)

weight_kg                                     classification  \
name                                                                      
Rattata          NaN                                      Mouse Pokémon   
Raticate         NaN                                      Mouse Pokémon   
Raichu           NaN                                      Mouse Pokémon   
Sandshrew        NaN                                      Mouse Pokémon   
Sandslash        NaN                                      Mouse Pokémon   
Vulpix           NaN                                        Fox Pokémon   
Ninetales        NaN                                        Fox Pokémon   
Diglett          NaN                                       Mole Pokémon   
Dugtrio          NaN                                       Mole Pokémon   
Meowth           NaN                                Scratch Cat Pokémon   
Persian          NaN                                 Classy Cat Pokémon   
Geodude          NaN                                       Rock Pokémon   
Graveler         NaN                                       Rock Pokémon   
Golem            NaN                                    Megaton Pokémon   
Grimer           NaN                                     Sludge Pokémon   
Muk              NaN                                     Sludge Pokémon   
Exeggutor        NaN                                    Coconut Pokémon   
Marowak          NaN                                Bone Keeper Pokémon   
Hoopa            NaN  Mischief Pokémon (Confined)Djinn Pokémonn (Unb...   
Lycanroc         NaN                                       Wolf Pokémon   

              type1     type2       name  
name                                      
Rattata      normal      dark    Rattata  
Raticate     normal      dark   Raticate  
Raichu     electric  electric     Raichu  
Sandshrew    ground       ice  Sandshrew  
Sandslash    ground       ice  Sandslash  
Vulpix         fire       ice     Vulpix  
Ninetales      fire       ice  Ninetales  
Diglett      ground    ground    Diglett  
Dugtrio      ground    ground    Dugtrio  
Meowth       normal      dark     Meowth  
Persian      normal      dark    Persian  
Geodude        rock    ground    Geodude  
Graveler       rock    ground   Graveler  
Golem          rock    ground      Golem  
Grimer       poison    poison     Grimer  
Muk          poison    poison        Muk  
Exeggutor     grass   psychic  Exeggutor  
Marowak      ground      fire    Marowak  
Hoopa       psychic     ghost      Hoopa  
Lycanroc       rock       NaN   Lycanroc

### Let's "double click" on *percentage_male*

In [ ]:
df_percentage_male_1 = df[ df['percentage_male'].isna() == 1 ]
df_percentage_male_1[ ['percentage_male','classification','type1','type2','name'] ].head(n=50)

percentage_male       classification     type1    type2       name
name                                                                         
Magnemite              NaN       Magnet Pokémon  electric    steel  Magnemite
Magneton               NaN       Magnet Pokémon  electric    steel   Magneton
Voltorb                NaN         Ball Pokémon  electric      NaN    Voltorb
Electrode              NaN         Ball Pokémon  electric      NaN  Electrode
Staryu                 NaN    Starshape Pokémon     water      NaN     Staryu
Starmie                NaN   Mysterious Pokémon     water  psychic    Starmie
Ditto                  NaN    Transform Pokémon    normal      NaN      Ditto
Porygon                NaN      Virtual Pokémon    normal      NaN    Porygon
Articuno               NaN       Freeze Pokémon       ice   flying   Articuno
Zapdos                 NaN     Electric Pokémon  electric   flying     Zapdos
Moltres                NaN        Flame Pokémon      fire   flying    Moltres
Mewtwo                 NaN      Genetic Pokémon   psychic      NaN     Mewtwo
Mew                    NaN  New Species Pokémon   psychic      NaN        Mew
Unown                  NaN       Symbol Pokémon   psychic      NaN      Unown
Porygon2               NaN      Virtual Pokémon    normal      NaN   Porygon2
Raikou                 NaN      Thunder Pokémon  electric      NaN     Raikou
Entei                  NaN      Volcano Pokémon      fire      NaN      Entei
Suicune                NaN       Aurora Pokémon     water      NaN    Suicune
Lugia                  NaN       Diving Pokémon   psychic   flying      Lugia
Ho-Oh                  NaN      Rainbow Pokémon      fire   flying      Ho-Oh
Celebi                 NaN  Time Travel Pokémon   psychic    grass     Celebi
Shedinja               NaN         Shed Pokémon       bug    ghost   Shedinja
Lunatone               NaN    Meteorite Pokémon      rock  psychic   Lunatone
Solrock                NaN    Meteorite Pokémon      rock  psychic    Solrock
Baltoy                 NaN    Clay Doll Pokémon    ground  psychic     Baltoy
Claydol                NaN    Clay Doll Pokémon    ground  psychic    Claydol
Beldum                 NaN    Iron Ball Pokémon     steel  psychic     Beldum
Metang                 NaN    Iron Claw Pokémon     steel  psychic     Metang
Metagross              NaN     Iron Leg Pokémon     steel  psychic  Metagross
Regirock               NaN    Rock Peak Pokémon      rock      NaN   Regirock
Regice                 NaN      Iceberg Pokémon       ice      NaN     Regice
Registeel              NaN         Iron Pokémon     steel      NaN  Registeel
Kyogre                 NaN    Sea Basin Pokémon     water      NaN     Kyogre
Groudon                NaN    Continent Pokémon    ground      NaN    Groudon
Rayquaza               NaN     Sky High Pokémon    dragon   flying   Rayquaza
Jirachi                NaN         Wish Pokémon     steel  psychic    Jirachi
Deoxys                 NaN          DNA Pokémon   psychic      NaN     Deoxys
Bronzor                NaN       Bronze Pokémon     steel  psychic    Bronzor
Bronzong               NaN  Bronze Bell Pokémon     steel  psychic   Bronzong
Magnezone              NaN  Magnet Area Pokémon  electric    steel  Magnezone
Porygon-Z              NaN      Virtual Pokémon    normal      NaN  Porygon-Z
Rotom                  NaN       Plasma Pokémon  electric    ghost      Rotom
Uxie                   NaN    Knowledge Pokémon   psychic      NaN       Uxie
Mesprit                NaN      Emotion Pokémon   psychic      NaN    Mesprit
Azelf                  NaN    Willpower Pokémon   psychic      NaN      Azelf
Dialga                 NaN     Temporal Pokémon     steel   dragon     Dialga
Palkia                 NaN      Spatial Pokémon     water   dragon     Palkia
Regigigas              NaN     Colossal Pokémon    normal      NaN  Regigigas
Giratina               NaN     Renegade Pokémon     ghost   dragon   Giratina
Phione                 NaN  Sea D

### How many generations do we have?

In [ ]:
df.drop_duplicates(subset=['generation'])

generation  is_legendary    type1   type2       classification  \
name                                                                        
Bulbasaur           1             0    grass  poison         Seed Pokémon   
Chikorita           2             0    grass     NaN         Leaf Pokémon   
Treecko             3             0    grass     NaN   Wood Gecko Pokémon   
Turtwig             4             0    grass     NaN    Tiny Leaf Pokémon   
Victini             5             1  psychic    fire      Victory Pokémon   
Chespin             6             0    grass     NaN    Spiky Nut Pokémon   
Rowlet              7             0    grass  flying  Grass Quill Pokémon   

                name                    abilities  against_bug  against_dark  \
name                                                                           
Bulbasaur  Bulbasaur  ['Overgrow', 'Chlorophyll']        1.000         1.000   
Chikorita  Chikorita   ['Overgrow', 'Leaf Guard']        2.000         1.000   
Treecko      Treecko     ['Overgrow', 'Unburden']        2.000         1.000   
Turtwig      Turtwig  ['Overgrow', 'Shell Armor']        2.000         1.000   
Victini      Victini             ['Victory Star']        1.000         2.000   
Chespin      Chespin  ['Overgrow', 'Bulletproof']        2.000         1.000   
Rowlet        Rowlet   ['Overgrow', 'Long Reach']        1.000         1.000   

           against_dragon  against_electric  against_fairy  against_fight  \
name                                                                        
Bulbasaur           1.000             0.500          0.500          0.500   
Chikorita           1.000             0.500          1.000          1.000   
Treecko             1.000             0.500          1.000          1.000   
Turtwig             1.000             0.500          1.000          1.000   
Victini             1.000             1.000          0.500          0.500   
Chespin             1.000             0.500          1.000          1.000   
Rowlet              1.000             1.000          1.000          0.500   

           against_fire  against_flying  against_ghost  against_grass  \
name                                                                    
Bulbasaur         2.000           2.000          1.000          0.250   
Chikorita         2.000           2.000          1.000          0.500   
Treecko           2.000           2.000          1.000          0.500   
Turtwig           2.000           2.000          1.000          0.500   
Victini           0.500           1.000          2.000          0.500   
Chespin           2.000           2.000          1.000          0.500   
Rowlet            2.000           2.000          1.000          0.250   

           against_ground  against_ice  against_normal  against_poison  \
name                                                                     
Bulbasaur           1.000        2.000           1.000           1.000   
Chikorita           0.500        2.000           1.000           2.000   
Treecko             0.500        2.000           1.000           2.000   
Turtwig             0.500        2.000           1.000           2.000   
Victini             2.000        0.500           1.000           1.000   
Chespin             0.500        2.000           1.000           2.000   
Rowlet              0.000        4.000           1.000           2.000   

           against_psychic  against_rock  against_steel  against_water  \
name                                                                     
Bulbasaur            2.000         1.000          1.000          0.500   
Chikorita            1.000         1.000          1.000          0.500   
Treecko              1.000         1.000          1.000          0.500   
Turtwig              1.000         1.000          1.000          0.500   
Victini              0.500         2.000          0.500          2.000   
Chespin              1.000         1.000          1.000    

In [ ]:
df.drop_duplicates(subset=['generation']).shape[0]

7

### How many pokemons are there in each generation?

In [ ]:
df[['generation']].value_counts()

generation
5             156
1             151
3             135
4             107
2             100
7              80
6              72
dtype: int64

In [ ]:
df[['generation']].value_counts().sort_index()

generation
1             151
2             100
3             135
4             107
5             156
6              72
7              80
dtype: int64

### How many pokemons are there per type in each generation?

In [ ]:
df_generation_type1 = df.groupby(['generation','type1'])
type(df_generation_type1)

pandas.core.groupby.generic.DataFrameGroupBy

In [ ]:
df_generation_type1.groups

{(1, 'bug'): ['Caterpie', 'Metapod', 'Butterfree', 'Weedle', 'Kakuna', 'Beedrill', 'Paras', 'Parasect', 'Venonat', 'Venomoth', 'Scyther', 'Pinsir'], (1, 'dragon'): ['Dratini', 'Dragonair', 'Dragonite'], (1, 'electric'): ['Pikachu', 'Raichu', 'Magnemite', 'Magneton', 'Voltorb', 'Electrode', 'Electabuzz', 'Jolteon', 'Zapdos'], (1, 'fairy'): ['Clefairy', 'Clefable'], (1, 'fighting'): ['Mankey', 'Primeape', 'Machop', 'Machoke', 'Machamp', 'Hitmonlee', 'Hitmonchan'], (1, 'fire'): ['Charmander', 'Charmeleon', 'Charizard', 'Vulpix', 'Ninetales', 'Growlithe', 'Arcanine', 'Ponyta', 'Rapidash', 'Magmar', 'Flareon', 'Moltres'], (1, 'ghost'): ['Gastly', 'Haunter', 'Gengar'], (1, 'grass'): ['Bulbasaur', 'Ivysaur', 'Venusaur', 'Oddish', 'Gloom', 'Vileplume', 'Bellsprout', 'Weepinbell', 'Victreebel', 'Exeggcute', 'Exeggutor', 'Tangela'], (1, 'ground'): ['Sandshrew', 'Sandslash', 'Diglett', 'Dugtrio', 'Cubone', 'Marowak', 'Rhyhorn', 'Rhydon'], (1, 'ice'): ['Jynx', 'Articuno'], (1, 'normal'): ['Pidgey'

In [ ]:
type(df_generation_type1.groups)

pandas.io.formats.printing.PrettyDict

In [ ]:
df_generation_type1.nunique()

is_legendary  type2  classification  name  abilities  \
generation type1                                                            
1          bug                  1      3              10    12          9   
           dragon               1      1               1     3          2   
           electric             2      3               5     9          7   
           fairy                1      0               1     2          2   
           fighting             1      0               4     7          4   
...                           ...    ...             ...   ...        ...   
7          poison               1      2               2     4          2   
           psychic              1      3               6     6          6   
           rock                 2      2               4     4          4   
           steel                1      2               2     2          2   
           water                2      3               8     9          6   

                     against_bug  against_dark  against_dragon  \
generation type1                                                 
1          bug                 3             1               1   
           dragon              2             1               1   
           electric            2             1               2   
           fairy               1             1               1   
           fighting            1             1               1   
...                          ...           ...             ...   
7          poison              2             1               1   
           psychic             2             3               3   
           rock                2             1               1   
           steel               1             2               2   
           water               3             3               2   

                     against_electric  against_fairy  against_fight  \
generation type1                                                      
1          bug                      3              2              2   
           dragon                   2              1              2   
           electric                 2              2              3   
           fairy                    1              1              1   
           fighting                 1              1              1   
...                               ...            ...            ...   
7          poison                   2              2              1   
           psychic                  1              2              4   
           rock                     2              2              2   
           steel                    2              1              1   
           water                    1              1              2   

                     against_fire  against_flying  against_ghost  \
generation type1                                                   
1          bug                  2               2              1   
           dragon               1               1              1   
           electric             2               2              1   
           fairy                1               1              1   
           fighting             1               1              1   
...                           ...             ...            ...   
7          poison               1               1              1   
           psychic              2               2              2   
           rock                 1               1              1   
           steel                1               1              1   
           water                2               2              2   

                     against_grass  against_ground  against_ice  \
generation type1                                                  
1          bug                   2               4            2   
           dragon                2               2            2   
           electric              2               3            3   
           fairy

In [ ]:
df_generation_type1.nunique()['name'].head(20)

generation  type1   
1           bug         12
            dragon       3
            electric     9
            fairy        2
            fighting     7
            fire        12
            ghost        3
            grass       12
            ground       8
            ice          2
            normal      22
            poison      14
            psychic      8
            rock         9
            water       28
2           bug         10
            dark         5
            electric     6
            fairy        5
            fighting     2
Name: name, dtype: int64

### How to reshape the layout of tables?

In [ ]:
df.pivot_table(index=['generation'], columns=['type1'], values="attack")

type1         bug    dark  dragon  electric  fairy  fighting   fire  flying  \
generation                                                                    
1          71.250     NaN  94.000    61.444 57.500   102.857 84.833     NaN   
2          77.000  79.000     NaN    63.000 57.000    65.000 76.250     NaN   
3          55.833  95.000 119.286    52.500    NaN    80.000 95.000     NaN   
4          61.000 111.667 110.000    81.857 50.000   107.500 85.000     NaN   
5          78.167  81.385 123.000    85.000    NaN   108.571 80.875 100.000   
6          36.333  92.333  81.250    50.333 68.444    99.333 69.375  50.000   
7          83.333 101.000  80.000   100.667 52.000   111.333 82.600     NaN   

type1       ghost  grass  ground    ice  normal  poison  psychic    rock  \
generation                                                                 
1          50.000 73.000  84.375 67.500  68.864  75.143   65.125  85.556   
2          60.000 57.000  85.000 58.750  63.133  90.000   69.286 103.000   
3          87.500 77.333  93.333 66.667  69.889  72.000   47.875  79.625   
4          83.667 82.154 104.750 90.000  83.471  77.167   85.000  86.500   
5          41.000 65.333 100.000 76.667  85.059  72.500   52.500 116.000   
6          86.500 82.200     NaN 93.000  55.500  67.500   85.333  89.125   
7          87.750 81.500 112.500    NaN  90.417  56.000   83.333  83.250   

type1        steel  water  
generation                 
1              NaN 72.036  
2          102.500 68.111  
3           95.000 77.000  
4           77.667 72.462  
5           81.250 73.353  
6          105.000 78.000  
7           98.000 76.333

In [ ]:
df.pivot_table(index=['type1'], columns=['generation'], values="attack")

generation       1       2       3       4       5       6       7
type1                                                             
bug         71.250  77.000  55.833  61.000  78.167  36.333  83.333
dark           NaN  79.000  95.000 111.667  81.385  92.333 101.000
dragon      94.000     NaN 119.286 110.000 123.000  81.250  80.000
electric    61.444  63.000  52.500  81.857  85.000  50.333 100.667
fairy       57.500  57.000     NaN  50.000     NaN  68.444  52.000
fighting   102.857  65.000  80.000 107.500 108.571  99.333 111.333
fire        84.833  76.250  95.000  85.000  80.875  69.375  82.600
flying         NaN     NaN     NaN     NaN 100.000  50.000     NaN
ghost       50.000  60.000  87.500  83.667  41.000  86.500  87.750
grass       73.000  57.000  77.333  82.154  65.333  82.200  81.500
ground      84.375  85.000  93.333 104.750 100.000     NaN 112.500
ice         67.500  58.750  66.667  90.000  76.667  93.000     NaN
normal      68.864  63.133  69.889  83.471  85.059  55.500  90.417
poison      75.143  90.000  72.000  77.167  72.500  67.500  56.000
psychic     65.125  69.286  47.875  85.000  52.500  85.333  83.333
rock        85.556 103.000  79.625  86.500 116.000  89.125  83.250
steel          NaN 102.500  95.000  77.667  81.250 105.000  98.000
water       72.036  68.111  77.000  72.462  73.353  78.000  76.333

In [ ]:
df.pivot_table(index=['type1','type2'], columns=['generation'], values="attack").head(20)

generation          1       2      3       4       5       6       7
type1 type2                                                         
bug   electric    NaN     NaN    NaN     NaN  62.000     NaN  76.000
      fairy       NaN     NaN    NaN     NaN     NaN     NaN  50.000
      fighting    NaN 185.000    NaN     NaN     NaN     NaN 138.000
      fire        NaN     NaN    NaN     NaN  72.500     NaN     NaN
      flying   77.500  40.000 73.333  70.000     NaN  52.000     NaN
      ghost       NaN     NaN 90.000     NaN     NaN     NaN     NaN
      grass    82.500     NaN    NaN  69.000  73.000     NaN     NaN
      ground      NaN     NaN 45.000     NaN     NaN     NaN     NaN
      poison   66.000  75.000 50.000     NaN  66.667     NaN     NaN
      rock        NaN  10.000    NaN     NaN  85.000     NaN     NaN
      steel       NaN 120.000    NaN     NaN 121.333     NaN     NaN
      water       NaN     NaN 30.000     NaN     NaN     NaN  80.000
dark  dragon      NaN     NaN    NaN     NaN  85.000     NaN 101.000
      fighting    NaN     NaN    NaN     NaN  82.500     NaN     NaN
      fire        NaN  75.000    NaN     NaN     NaN     NaN     NaN
      flying      NaN  85.000    NaN 125.000  60.000 131.000     NaN
      ghost       NaN     NaN 85.000     NaN     NaN     NaN     NaN
      ice         NaN  95.000    NaN 120.000     NaN     NaN     NaN
      psychic     NaN     NaN    NaN     NaN     NaN  73.000     NaN
      steel       NaN     NaN    NaN     NaN 105.000     NaN     NaN

In [ ]:
df.melt( 
        id_vars=['name'], 
        value_vars=['is_legendary','attack','defense','speed'],
        var_name='Dimension', 
        value_name='Metric'
).sort_values(by=['name','Dimension'] ).head(20)

name     Dimension  Metric
1260  Abomasnow        attack     132
2061  Abomasnow       defense     105
459   Abomasnow  is_legendary       0
2862  Abomasnow         speed      30
863        Abra        attack      20
1664       Abra       defense      15
62         Abra  is_legendary       0
2465       Abra         speed      90
1159      Absol        attack     150
1960      Absol       defense      60
358       Absol  is_legendary       0
2761      Absol         speed     115
1417   Accelgor        attack      70
2218   Accelgor       defense      40
616    Accelgor  is_legendary       0
3019   Accelgor         speed     145
1481  Aegislash        attack     150
2282  Aegislash       defense      50
680   Aegislash  is_legendary       0
3083  Aegislash         speed      60

## Further reading

- [Comparison with spreadsheets](https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_spreadsheets.html#compare-with-spreadsheets)
- [Comparison with SQL](https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_sql.html#compare-with-sql)
- [Comparison with R / R libraries](https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_r.html#compare-with-r)
- [Comparison with Stata](https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_stata.html#compare-with-stata)
- [Comparison with SAS](https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_sas.html#compare-with-sas)
- [How do I create plots in pandas?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/04_plotting.html)
- [How to combine data from multiple tables?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/08_combine_dataframes.html)
- [How to handle time series data with ease?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/09_timeseries.html)
- [How to manipulate textual data?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/10_text_data.html#min-tut-10-text)

---
If you made it this far, follow [David Regalado](https://beacons.ai/davidregalado) for more code!